auditlog/models.py

```diff
@@ -208,6 +216,80 @@ def _get_pk_value(self, instance):
             pk = self._get_pk_value(pk)
         return pk
 
+    def _get_serialized_data_or_none(self, instance):
+        from .registry import auditlog
+
+        opts = auditlog.get_serialize_options(instance.__class__)
+        if not opts["serialize_data"]:
+            return None
+
+        model_fields = auditlog.get_model_fields(instance.__class__)
+        kwargs = opts.get("serialize_kwargs", {})
+
+        if opts["serialize_auditlog_fields_only"]:
+            kwargs.setdefault(
+                "fields", self._get_applicable_model_fields(instance, model_fields)
+            )
+
+        instance_copy = self._get_copy_with_python_typed_fields(instance)
+        data = dict(
+            json.loads(serializers.serialize("json", (instance_copy,), **kwargs))[0]
+        )
+
+        mask_fields = model_fields["mask_fields"]
+        if mask_fields:
+            data = self._mask_serialized_fields(data, mask_fields)
+
+        return data
+
+    def _get_copy_with_python_typed_fields(self, instance):
+        """
+        Attempt to create copy of instance and coerce types on instance fields
+
+        The Django core serializer assumes that the values on object fields are
+        correctly typed to their respective fields. Updates made to an object's
+        in-memory state may not meet this assumption. To prevent this violation, values
+        are typed by calling `to_python` from the field object, the result is set on a
+        copy of the instance and the copy is sent to the serializer.
+        """
+        try:
+            instance_copy = deepcopy(instance)
+        except TypeError:
+            instance_copy = instance
+        for field in instance_copy._meta.fields:
+            value = getattr(instance_copy, field.name)
+            if not issubclass(type(value), models.Model):
+                setattr(instance_copy, field.name, field.to_python(value))
+        return instance_copy
+
+    def _get_applicable_model_fields(
+        self, instance, model_fields: Dict[str, List[str]]
+    ) -> List[str]:
+        include_fields = model_fields["include_fields"]
+        exclude_fields = model_fields["exclude_fields"]
+        all_field_names = [field.name for field in instance._meta.fields]
+
+        if not include_fields and not exclude_fields:
+            return all_field_names
+
+        return list(set(include_fields or all_field_names).difference(exclude_fields))
+
+    def _mask_serialized_fields(
+        self, data: Dict[str, Any], mask_fields: List[str]
+    ) -> Dict[str, Any]:
+        masked_data = data.copy()
+        all_field_data = dict(masked_data.pop("fields"))
```

Comment:
Do we really need to convert the `fields` content to dict?

Reply:
My thought was that we might want to attempt to set the type and error here rather than sending an unknown type to  the iterator. I suppose it doesn't actually matter where the TypeError would happen. And I'm not aware of any case where the serializer would return a different shape without the nested fields dictionary either. I've removed the conversion.